Adds long compression methods

acslk opened the pull request:
This PR introduces a few new compression strategies for long values that are not block based (described in #3147).  The enum CompressionFormat is added to control the compression procedure, which could call GenericIndexed for block based compression.

The compression format added include delta compression, table compression, and a new uncompressed format. Delta compression finds the smallest value in the segment, and store all values as offsets. Table compression maps unique values and store their id. The offsets and ids are stored by using the smallest number of bits required for the maximum values. For example, if the maximum offset is 200, then all values are stored using 8 bits. The new uncompressed format is the uncompressed format without the header for position of each block and empty bytes at beginning of each block.

Update

After some discussion in #3147 , it's decided that compression strategy, which compress and decompress blocks of data without knowing the data format, and encoding format, which encode data using knowledge of its attributes such as size of each element, should be separated. Users would choose the combination of compression and encoding to use.

The existing compressions including LZF, LZ4, and Uncompressed all count as compression strategy. In addition, this PR included the compression NONE, which indicate no compression strategy should be used, and the values should be stored all together instead of in blocks, since this would allow faster random access. The LZF and Uncompressed strategy are still needed for backward compatibility, but they shouldn't be used for new data.

The encoding format introduced in this PR are Delta, Table, and Longs. Delta and table are described above, while Longs just indicated values are stored as 8 byte longs. Segments created prior to this PR are viewed as compression strategy used + Longs encoding. The strategy for choosing the encoding are auto and longs. Auto strategy scans through the data for its cardinality and maximum offset, and choose the most appropriate format. Longs strategy just always choose longs format.

Below is a benchmark of combinations of all compression strategies and encoding strategies using some generated data sets.

Data sets includes: 
- enumerated (value from only a few selections, with probability heavily skewed toward one) 
- zipfLow (zipf distribution between 0 to 1000 with low exponent)
- zipfHigh (zipf distribution between 0 to 1000 with high exponent)
- uniform (uniform distribution between 0 to 1000)
- sequential (sequential values starting from a timestamp)

Generated files size in kb (5,000,000 values)
```
zipfLow-LZ4-DELTA: 6870
zipfLow-LZ4-LONGS: 9495
zipfLow-NONE-DELTA: 7324
zipfLow-NONE-LONGS: 39062
uniform-LZ4-DELTA: 7354
uniform-LZ4-LONGS: 11184
uniform-NONE-DELTA: 7324
uniform-NONE-LONGS: 39062
sequential-LZ4-DELTA: 14706
sequential-LZ4-LONGS: 19542
sequential-NONE-DELTA: 14648
sequential-NONE-LONGS: 39062
enumerate-LZ4-DELTA: 442
enumerate-LZ4-LONGS: 711
enumerate-NONE-DELTA: 2441
enumerate-NONE-LONGS: 39062
zipfHigh-LZ4-DELTA: 1201
zipfHigh-LZ4-LONGS: 1784
zipfHigh-NONE-DELTA: 4884
zipfHigh-NONE-LONGS: 39062
```

Benchmarks:
readContinuous reads every value, while readSkipping skips randomly between 0 - 2000 values on each read
```
Read Continuous
 enumerate      auto   lz4  30.587 ± 0.493  ms/op
 enumerate      auto  none   7.093 ± 0.116  ms/op
 enumerate     longs   lz4  19.731 ± 0.849  ms/op
 enumerate     longs  none   3.107 ± 0.060  ms/op
   zipfLow      auto   lz4  36.530 ± 0.695  ms/op
   zipfLow      auto  none  11.070 ± 0.209  ms/op
   zipfLow     longs   lz4  33.192 ± 0.652  ms/op
   zipfLow     longs  none   3.127 ± 0.076  ms/op
  zipfHigh      auto   lz4  28.517 ± 0.804  ms/op
  zipfHigh      auto  none   2.808 ± 0.072  ms/op
  zipfHigh     longs   lz4  21.903 ± 0.508  ms/op
  zipfHigh     longs  none   3.033 ± 0.040  ms/op
sequential      auto   lz4  29.984 ± 0.807  ms/op
sequential      auto  none   6.771 ± 0.090  ms/op
sequential     longs   lz4  54.404 ± 1.106  ms/op
sequential     longs  none   3.172 ± 0.088  ms/op
   uniform      auto   lz4  32.214 ± 0.803  ms/op
   uniform      auto  none  10.702 ± 0.145  ms/op
   uniform     longs   lz4  31.917 ± 0.663  ms/op
   uniform     longs  none   3.055 ± 0.072  ms/op

ReadSkipping
 enumerate      auto   lz4   1.533 ± 0.075  ms/op
 enumerate      auto  none   0.083 ± 0.002  ms/op
 enumerate     longs   lz4   5.862 ± 0.471  ms/op
 enumerate     longs  none   0.285 ± 0.003  ms/op
   zipfLow      auto   lz4   5.017 ± 0.121  ms/op
   zipfLow      auto  none   0.172 ± 0.003  ms/op
   zipfLow     longs   lz4  19.103 ± 1.118  ms/op
   zipfLow     longs  none   0.283 ± 0.002  ms/op
  zipfHigh      auto   lz4   3.060 ± 0.164  ms/op
  zipfHigh      auto  none   0.109 ± 0.002  ms/op
  zipfHigh     longs   lz4   8.791 ± 1.083  ms/op
  zipfHigh     longs  none   0.280 ± 0.002  ms/op
sequential      auto   lz4   1.379 ± 0.304  ms/op
sequential      auto  none   0.253 ± 0.003  ms/op
sequential     longs   lz4  40.777 ± 0.681  ms/op
sequential     longs  none   0.281 ± 0.003  ms/op
   uniform      auto   lz4   0.822 ± 0.289  ms/op
   uniform      auto  none   0.171 ± 0.003  ms/op
   uniform     longs   lz4  18.799 ± 1.704  ms/op
   uniform     longs  none   0.286 ± 0.003  ms/op
```

Benchmark on master, lz4 correspond to longs + lz4. There is no major performance difference
```
ReadContinous
 enumerate     lz4  19.714 ± 0.455  ms/op
   zipfLow     lz4  33.191 ± 0.640  ms/op
  zipfHigh     lz4  21.761 ± 0.488  ms/op
sequential     lz4  54.163 ± 1.077  ms/op
   uniform     lz4  31.790 ± 0.871  ms/op

ReadSkipping
 enumerate     lz4   5.741 ± 0.346  ms/op
   zipfLow     lz4  18.736 ± 0.425  ms/op
  zipfHigh     lz4   8.739 ± 0.753  ms/op
sequential     lz4  41.081 ± 1.034  ms/op
   uniform     lz4  18.322 ± 0.325  ms/op
```

acslk commented:
After some discussion in #3147 , it's decided that _compression strategy_, which compress and decompress blocks of data without knowing the data format, and _encoding format_, which encode data using knowledge of its attributes such as size of each element, should be separated. Users would choose the combination of compression and encoding to use.

The existing compressions including LZF, LZ4, and Uncompressed all count as _compression strategy_. In addition, this PR included the compression NONE, which indicate no compression strategy should be used, and the values should be stored all together instead of in blocks, since this would allow faster random access. The LZF and Uncompressed strategy are still needed for backward compatibility, but they shouldn't be used for new data.

The _encoding format_ introduced in this PR are Delta, Table, and Longs. Delta and table are described in previous comment, while Longs just indicated values are stored as 8 byte longs. Segments created prior to this PR are viewed as compression strategy used + Longs encoding.

Below is a benchmark of combinations of all compression strategies and encoding formats
```
CompressionFormatBenchmark.readContinuous     reqInt     delta           lz4  avgt   10  29.286 ± 1.393  ms/op
CompressionFormatBenchmark.readContinuous     reqInt     delta  uncompressed  avgt   10  29.431 ± 1.787  ms/op
CompressionFormatBenchmark.readContinuous     reqInt     delta          none  avgt   10   8.065 ± 0.522  ms/op
CompressionFormatBenchmark.readContinuous     reqInt     longs           lz4  avgt   10  21.420 ± 3.677  ms/op
CompressionFormatBenchmark.readContinuous     reqInt     longs  uncompressed  avgt   10  18.791 ± 1.351  ms/op
CompressionFormatBenchmark.readContinuous     reqInt     longs          none  avgt   10   3.029 ± 0.076  ms/op
CompressionFormatBenchmark.readContinuous  bytesLong     delta           lz4  avgt   10  44.828 ± 4.710  ms/op
CompressionFormatBenchmark.readContinuous  bytesLong     delta  uncompressed  avgt   10  31.041 ± 2.123  ms/op
CompressionFormatBenchmark.readContinuous  bytesLong     delta          none  avgt   10   8.400 ± 1.480  ms/op
CompressionFormatBenchmark.readContinuous  bytesLong     longs           lz4  avgt   10  34.514 ± 4.275  ms/op
CompressionFormatBenchmark.readContinuous  bytesLong     longs  uncompressed  avgt   10  20.392 ± 3.440  ms/op
CompressionFormatBenchmark.readContinuous  bytesLong     longs          none  avgt   10   3.028 ± 0.049  ms/op
CompressionFormatBenchmark.readContinuous  timestamp     delta           lz4  avgt   10  36.941 ± 2.615  ms/op
CompressionFormatBenchmark.readContinuous  timestamp     delta  uncompressed  avgt   10  30.484 ± 2.581  ms/op
CompressionFormatBenchmark.readContinuous  timestamp     delta          none  avgt   10   3.366 ± 0.187  ms/op
CompressionFormatBenchmark.readContinuous  timestamp     longs           lz4  avgt   10  53.574 ± 6.133  ms/op
CompressionFormatBenchmark.readContinuous  timestamp     longs  uncompressed  avgt   10  19.124 ± 2.194  ms/op
CompressionFormatBenchmark.readContinuous  timestamp     longs          none  avgt   10   3.048 ± 0.061  ms/op
CompressionFormatBenchmark.readSkipping       reqInt     delta           lz4  avgt   10   0.740 ± 0.104  ms/op
CompressionFormatBenchmark.readSkipping       reqInt     delta  uncompressed  avgt   10   0.196 ± 0.122  ms/op
CompressionFormatBenchmark.readSkipping       reqInt     delta          none  avgt   10   0.017 ± 0.001  ms/op
CompressionFormatBenchmark.readSkipping       reqInt     longs           lz4  avgt   10   4.681 ± 3.324  ms/op
CompressionFormatBenchmark.readSkipping       reqInt     longs  uncompressed  avgt   10   3.909 ± 5.152  ms/op
CompressionFormatBenchmark.readSkipping       reqInt     longs          none  avgt   10   0.059 ± 0.001  ms/op
CompressionFormatBenchmark.readSkipping    bytesLong     delta           lz4  avgt   10  14.634 ± 1.457  ms/op
CompressionFormatBenchmark.readSkipping    bytesLong     delta  uncompressed  avgt   10   1.780 ± 3.046  ms/op
CompressionFormatBenchmark.readSkipping    bytesLong     delta          none  avgt   10   0.054 ± 0.003  ms/op
CompressionFormatBenchmark.readSkipping    bytesLong     longs           lz4  avgt   10  18.532 ± 2.257  ms/op
CompressionFormatBenchmark.readSkipping    bytesLong     longs  uncompressed  avgt   10   3.872 ± 4.596  ms/op
CompressionFormatBenchmark.readSkipping    bytesLong     longs          none  avgt   10   0.059 ± 0.001  ms/op
CompressionFormatBenchmark.readSkipping    timestamp     delta           lz4  avgt   10   7.756 ± 0.641  ms/op
CompressionFormatBenchmark.readSkipping    timestamp     delta  uncompressed  avgt   10   1.684 ± 0.804  ms/op
CompressionFormatBenchmark.readSkipping    timestamp     delta          none  avgt   10   0.056 ± 0.003  ms/op
CompressionFormatBenchmark.readSkipping    timestamp     longs           lz4  avgt   10  37.492 ± 4.476  ms/op
CompressionFormatBenchmark.readSkipping    timestamp     longs  uncompressed  avgt   10   3.377 ± 2.540  ms/op
CompressionFormatBenchmark.readSkipping    timestamp     longs          none  avgt   10   0.060 ± 0.002  ms/op
```

Size of compressed files
```
 9994113 bytesLong.delta-lz4
15976162 bytesLong.delta-none
16027498 bytesLong.delta-uncompressed
11119760 bytesLong.longs-lz4
42603034 bytesLong.longs-none
42669164 bytesLong.longs-uncompressed
  177951 reqInt.delta-lz4
 2662773 reqInt.delta-none
 2688050 reqInt.delta-uncompressed
  388076 reqInt.longs-lz4
42603034 reqInt.longs-none
42669164 reqInt.longs-uncompressed
19502946 timestamp.delta-lz4
21301540 timestamp.delta-none
21339814 timestamp.delta-uncompressed
19270814 timestamp.longs-lz4
42603034 timestamp.longs-none
42669164 timestamp.longs-uncompressed

```

drcrallen commented on processing/src/main/java/io/druid/segment/data/BlockLayoutSerde.java:
```diff
@@ -0,0 +1,309 @@
+    private final ByteOrder order;
+    private final CompressionFactory.LongEncodingFormatReader baseReader;
+
+    public BlockLayoutIndexedLongsSupplier(int totalSize, int sizePer, ByteBuffer fromBuffer, ByteOrder order,
```
Add javadoc here please

drcrallen commented on processing/src/main/java/io/druid/segment/data/BlockLayoutSerde.java:
```diff
@@ -0,0 +1,309 @@
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.WritableByteChannel;
+
+public class BlockLayoutSerde
```
Github is kind of hard to read. Is there anything actually in this class? if it is just a wrapper for the other classes can you just make it a package instead?

drcrallen commented on processing/src/main/java/io/druid/segment/data/BlockLayoutSerde.java:
```diff
@@ -0,0 +1,309 @@
+      }
+
+      @Override
+      public long get(int index)
```
can you just make this abstract and put all the get logic in the same general area?

acslk replied:
I'm not sure if that would be better, the idea behind the concrete base method is that this should be  the generic way of get, the overrides are all conditional optimizations. The optimization does not change the functionality of the method.

drcrallen commented on processing/src/main/java/io/druid/segment/data/BlockLayoutSerde.java:
```diff
@@ -0,0 +1,309 @@
+              )
+          );
+        }
+//        int bufferNum = index / sizePer;
```
remove or uncomment.

drcrallen commented on processing/src/main/java/io/druid/segment/data/BlockLayoutSerde.java:
```diff
@@ -0,0 +1,309 @@
+    }
+
+    public void close() throws IOException {
+      if (endBuffer != null) {
```
Is this intended to support multiple calls to close?

acslk replied:
The previous CompressedLongSupplierSerializer had these methods for close. It seems that IndexMerger use closeAndConsolidate() to combine the data, while IndexMergerV9 use close() followed by writeToChannel().

fjy commented on benchmarks/src/main/java/io/druid/benchmark/CompressionFormatBenchmark.java:
```diff
@@ -0,0 +1,144 @@
+package io.druid.benchmark;
```
header

fjy commented on benchmarks/src/main/java/io/druid/benchmark/VSizeSerdeBenchmark.java:
```diff
@@ -0,0 +1,181 @@
+package io.druid.benchmark;
```
header

fjy commented on processing/src/main/java/io/druid/segment/IndexSpec.java:
```diff
@@ -91,7 +94,8 @@ public IndexSpec()
       @JsonProperty("dimensionCompression") String dimensionCompression,
-      @JsonProperty("metricCompression") String metricCompression
+      @JsonProperty("metricCompression") String metricCompression,
+      @JsonProperty("metricLongEncoding") String metricLongEncoding
```
this will require documentation updates.

In general we need docs for all these new compression methods

xvrl replied:
can we call this metricEncoding to keep the configuration simple? We can simply return an error if someone uses an encoding that doesn't apply to a certain type of metric.

xvrl replied:
can we also define constants somewhere for acceptable encodings?

acslk replied:
@xvrl I think it's a good idea to have a separate set of encodings for each data type since encoding is generally not applicable to multiple data types. If this is the case, it doesn't make much sense to have a common metricEncoding value since no type specific encoding can be used.

For the acceptable encodings, I followed the same way as CompressionStrategy, where acceptable values are lowercase string of all LongEncodingFormat enum values.

gianm replied:
This is more like a `longEncoding`, assuming it will apply to long dimensions and `__time` too.

fjy commented on processing/src/main/java/io/druid/segment/data/BlockLayoutSerde.java:
```diff
@@ -0,0 +1,309 @@
+package io.druid.segment.data;
```
header

fjy commented on processing/src/main/java/io/druid/segment/data/BlockUncompressedFormatSerde.java:
```diff
@@ -0,0 +1,91 @@
+package io.druid.segment.data;
```
header

fjy commented on processing/src/main/java/io/druid/segment/data/CompressionFactory.java:
```diff
@@ -0,0 +1,186 @@
+package io.druid.segment.data;
```
header

fjy commented on processing/src/main/java/io/druid/segment/data/DeltaEncodingFormatSerde.java:
```diff
@@ -0,0 +1,133 @@
+package io.druid.segment.data;
```
header

fjy commented on processing/src/main/java/io/druid/segment/data/EntireLayoutSerde.java:
```diff
@@ -0,0 +1,160 @@
+package io.druid.segment.data;
```
header

fjy commented on processing/src/main/java/io/druid/segment/data/IntermediateSerde.java:
```diff
@@ -0,0 +1,133 @@
+package io.druid.segment.data;
```
header

fjy commented on processing/src/main/java/io/druid/segment/data/LongsEncodingFormatSerde.java:
```diff
@@ -0,0 +1,125 @@
+package io.druid.segment.data;
```
header

xvrl commented on benchmarks/src/main/java/io/druid/benchmark/CompressionFormatBenchmark.java:
```diff
@@ -0,0 +1,144 @@
+  @Param({"reqInt", "bytesLong", "timestamp"})
+  private static String file;
+
+  @Param({"delta", "longs"})
```
we should define those as constants

fjy commented on processing/src/main/java/io/druid/segment/data/TableEncodingFormatSerde.java:
```diff
@@ -0,0 +1,150 @@
+package io.druid.segment.data;
```
header

fjy commented on processing/src/main/java/io/druid/segment/data/VSizeCompressedObjectStrategy.java:
```diff
@@ -0,0 +1,63 @@
+package io.druid.segment.data;
```
header

fjy commented on processing/src/main/java/io/druid/segment/data/VSizeLongSerde.java:
```diff
@@ -0,0 +1,589 @@
+package io.druid.segment.data;
```
header

gianm commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -743,12 +745,13 @@ private LongColumnSerializer setupTimeWriter(final IOPeon ioPeon) throws IOExcep
   {
     ArrayList<GenericColumnSerializer> metWriters = Lists.newArrayListWithCapacity(mergedMetrics.size());
     final CompressedObjectStrategy.CompressionStrategy metCompression = indexSpec.getMetricCompressionStrategy();
+    final CompressionFactory.LongEncodingFormat metEncoding = indexSpec.getMetricLongEncodingFormat();
```
This is more like a `longEncoding` than a `metEncoding` (keep in mind we will have long dimensions soon; also, `__time` is a long).

gianm commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -726,7 +727,8 @@ private void mergeIndexesAndWriteColumns(
     LongColumnSerializer timeWriter = LongColumnSerializer.create(
-        ioPeon, "little_end_time", CompressedObjectStrategy.DEFAULT_COMPRESSION_STRATEGY
+        ioPeon, "little_end_time", CompressedObjectStrategy.DEFAULT_COMPRESSION_STRATEGY,
+        CompressionFactory.DEFAULT_LONG_ENCODING
```
`__time` should still write with with the longEncoding from indexSpec.

acslk replied:
time uses the default compression strategy and isn't set in indexSpec, so I'm not sure whether to do the same for encoding considering time is a special column

gianm replied:
I think it's fine that it does its own compression thing, as those configs are called dimensionCompression and metricCompression, so, okay, fine, it's not clear which one should apply to __time (it behaves like a dimension but is not stored the same way as "classic" string dimensions).

But __time is definitely a long column, that's for sure, so `longEncoding` should apply to it.

acslk replied:
changed timeWriter to use long encoding from indexSpec

gianm commented on processing/src/main/java/io/druid/segment/IndexMergerV9.java:
```diff
@@ -743,12 +745,13 @@ private LongColumnSerializer setupTimeWriter(final IOPeon ioPeon) throws IOExcep
   {
     ArrayList<GenericColumnSerializer> metWriters = Lists.newArrayListWithCapacity(mergedMetrics.size());
     final CompressedObjectStrategy.CompressionStrategy metCompression = indexSpec.getMetricCompressionStrategy();
+    final CompressionFactory.LongEncodingFormat metEncoding = indexSpec.getLongEncodingFormat();
```
Would better be called `longEncoding` than `metEncoding`.

acslk replied:
yeah left this out when renaming everything, changed now

gianm commented on processing/src/main/java/io/druid/segment/data/CompressionFactory.java:
```diff
@@ -0,0 +1,247 @@
+   * There is no header or version for Longs encoding format for backward compatibility
+   */
+
+  public enum LongEncodingFormat
```
Would better be called `LongEncoding` instead of `LongEncodingFormat`, IMO

acslk replied:
Renamed LongEncodingFormat to LongEncoding, as well as LongEncodingFormatReader and LongEncodingFormatWriter

gianm commented on processing/src/main/java/io/druid/segment/IndexSpec.java:
```diff
@@ -123,6 +132,12 @@ public String getMetricCompression()
     return metricCompression;
   }
 
+  @JsonProperty("longEncoding")
```
No need for `@JsonProperty` here, "longEncoding" would be the default.

acslk replied:
removed ("longEncoding") part

gianm commented on processing/src/main/java/io/druid/segment/IndexMerger.java:
```diff
@@ -763,8 +766,9 @@ public void close() throws IOException
-          ioPeon, "little_end_time", IndexIO.BYTE_ORDER, CompressedObjectStrategy.DEFAULT_COMPRESSION_STRATEGY
+      LongSupplierSerializer timeWriter = CompressionFactory.getLongSerializer(
+          ioPeon, "little_end_time", IndexIO.BYTE_ORDER, CompressionFactory.DEFAULT_LONG_ENCODING,
+          CompressedObjectStrategy.DEFAULT_COMPRESSION_STRATEGY
```
`__time` should write with with the longEncoding from indexSpec.

acslk replied:
changed timeWriter to use long encoding form indexSpec

jon-wei commented on processing/src/main/java/io/druid/segment/data/CompressedObjectStrategy.java:
```diff
@@ -44,6 +44,8 @@
   public static final CompressionStrategy DEFAULT_COMPRESSION_STRATEGY = CompressionStrategy.LZ4;
 
+  // When adding compression strategy do not use id in the range [0x7C, 0xFD], since the id could be subtracted by 126
+  // to indicate whether an encoding format id is needed
```
Can you add more detail on the reason for subtracting 126?

acslk replied:
moved all subtracting 126 and compare 0xFE code to CompressionFactory class with some explanation on the flag setting

jon-wei commented on docs/content/ingestion/tasks.md:
```diff
@@ -122,8 +122,9 @@ compression formats. The indexSpec is optional and default parameters will be us
 |bitmap|Object|Compression format for bitmap indexes. Should be a JSON object; see below for options.|no (defaults to Concise)|
-|dimensionCompression|String|Compression format for dimension columns. Choose from `LZ4`, `LZF`, or `uncompressed`.|no (default == `LZ4`)|
-|metricCompression|String|Compression format for metric columns. Choose from `LZ4`, `LZF`, or `uncompressed`.|no (default == `LZ4`)|
+|dimensionCompression|String|Compression format for dimension columns. Choose from `LZ4`, `LZF`, `uncompressed`, or `none`. Note that `uncompressed` is deprecated and should only be used for backward compatibility, for newer versions, use `none` instead. |no (default == `LZ4`)|
```
should this and similar docs have a note on why "uncompressed" is being deprecated/difference between "uncompressed" and "none"?

acslk replied:
maybe it's a good idea to have a separate page talking about the metrics storage format, and move this note and the explanation there

jon-wei replied:
sure, that sounds good

jon-wei commented on processing/src/main/java/io/druid/segment/data/CompressedLongsIndexedSupplier.java:
```diff
@@ -73,302 +60,58 @@ public int size()
+        if (compressionId >= (byte) 0xFE) {
+          compression = CompressedObjectStrategy.CompressionStrategy.forId(compressionId);
+        } else {
+          compression = CompressedObjectStrategy.CompressionStrategy.forId((byte) (compressionId + 126));
```
can we make 126 a constant value (maybe in CompressionStrategy?)

jon-wei commented on processing/src/main/java/io/druid/segment/data/CompressedLongsIndexedSupplier.java:
```diff
@@ -73,302 +60,58 @@ public int size()
+      CompressionFactory.LongEncodingFormat encoding = CompressionFactory.DEFAULT_LONG_ENCODING;
+      if (versionFromBuffer == version) {
+        byte compressionId = bufferToUse.get();
+        if (compressionId >= (byte) 0xFE) {
```
can we make this >= 0xFE check a method, maybe something like compressionHeaderHasEncoding()

jon-wei commented on processing/src/main/java/io/druid/segment/data/VSizeLongSerde.java:
```diff
@@ -0,0 +1,633 @@
+    @Override
+    public void write(long value) throws IOException
+    {
+      int shift = 0;
```
is `shift` variable needed? Seems like it gets set to 4 once before it's used and never changes after

jon-wei commented on benchmarks/src/main/java/io/druid/benchmark/VSizeSerdeBenchmark.java:
```diff
@@ -0,0 +1,199 @@
+    buffer = Files.map(new File(this.getClass().getClassLoader().getResource("timestamp.out").toURI()));
+    d1 = VSizeLongSerde.getDeserializer(1, buffer, 10);
+    d2 = VSizeLongSerde.getDeserializer(2, buffer, 10);
+    //d3 = VSizeLongSerde.getDeserializer(3, buffer, 10);
```
let's get rid of the unsued d3 and d6 deserializers

acslk replied:
removed unused code

jon-wei commented on processing/src/main/java/io/druid/segment/LongMetricColumnSerializer.java:
```diff
@@ -51,8 +54,8 @@ public LongMetricColumnSerializer(
-    writer = CompressedLongsSupplierSerializer.create(
-        ioPeon, String.format("%s_little", metricName), IndexIO.BYTE_ORDER,
+    writer = CompressionFactory.getLongSerializer(
+        ioPeon, String.format("%s_little", metricName), IndexIO.BYTE_ORDER, CompressionFactory.DEFAULT_LONG_ENCODING,
```
Should this use the user-specified long encoding instead of the default?

acslk replied:
previously segments created by using IndexMerger then V8toV9 conversion didn't use compression strategy specified by the indexSpec, so I also used the default encoding. Changed it in the newer commit to use compression and encoding from indexSpec instead

jon-wei commented on processing/src/main/java/io/druid/segment/data/VSizeLongSerde.java:
```diff
@@ -0,0 +1,633 @@
+  }
+
+
+  // LongDeserializers were adapted from Apache Lucene DirectReader, see:
```
let's add this to the NOTICE file

acslk replied:
added apache lucene use to NOTICE file

jon-wei commented on processing/src/main/java/io/druid/segment/data/LongsEncodingReader.java:
```diff
@@ -63,6 +60,6 @@ public int getNumBytes(int values)
-    return new LongsEncodingReader(buffer, order);
+    return new LongsEncodingReader(buffer.duplicate());
   }
 }
```
newline at EOF

acslk replied:
added newline

jon-wei commented on benchmarks/src/main/java/io/druid/benchmark/VSizeSerdeBenchmark.java:
```diff
@@ -96,103 +93,103 @@ public void tearDown()
   public void read1()
   {
-    for (int i = 0; i < 5000000; i++) {
+    for (int i = 0; i < 500000; i++) {
```
should this count be parameterized?

jon-wei commented on NOTICE:
```diff
@@ -16,3 +16,9 @@ This product contains conjunctive normal form conversion code adapted from Apach
+  * LICENSE:
+    * https://github.com/apache/lucene-solr/blob/master/lucene/LICENSE.txt (Apache License, Version 2.0)
+  * HOMEPAGE:
+    * https://github.com/apache/lucene-solr
```
newline EOF

acslk replied:
added newline

jon-wei commented on benchmarks/src/main/java/io/druid/benchmark/VSizeSerdeBenchmark.java:
```diff
@@ -0,0 +1,196 @@
+  @Benchmark
+  public void read1()
+  {
+    for (int i = 0; i < 500000; i++) {
```
should this count be parameterized?

acslk replied:
parameterized the value

jon-wei commented:
some minor comments, :+1: after those

gianm commented on benchmarks/src/main/java/io/druid/benchmark/CompressionBenchmark.java:
```diff
@@ -0,0 +1,168 @@
+public class CompressionBenchmark
+{
+  @Param({"reqInt", "bytesLong", "timestamp"})
+  private static String file;
```
Is it possible to get similar distributions using the BenchmarkDataGenerator? That'd mean we don't need these mondo files in the git repo. It'd also make it more clear what their distributions are supposed to be.

acslk replied:
Yeah that sounds like the better way, I'll try getting some similar results with the generator.

gianm replied:
Would also be nice to see storage "benchmarks" here too. i.e., a class that just has a main that generates/loads data with a particular distribution and prints out how much space the generated segments take up with:

- none + longs
- none + delta/table
- lz4 + longs
- lz4 + delta/table

gianm commented on processing/src/main/java/io/druid/segment/data/GenericIndexed.java:
```diff
@@ -47,7 +47,7 @@
 public class GenericIndexed<T> implements Indexed<T>
 {
-  private static final byte version = 0x1;
+  public static final byte version = 0x1;
```
This doesn't seem necessary

acslk replied:
changed it back to private

gianm commented on processing/src/main/java/io/druid/segment/data/CompressedObjectStrategy.java:
```diff
@@ -85,6 +92,22 @@ public Compressor getCompressor()
+      @Override
+      public Decompressor getDecompressor()
+      {
+        return UncompressedDecompressor.defaultDecompressor;
```
Nothing should call this, right? If not, let's do `throw new UnsupportedOperationException("No compression means no decompressor!")`

acslk replied:
Changed this to throw exception. Also modified some tests for compressed float and ints to not use NONE since they don't support properly reading NONE compression type. It makes sense since we never want to use uncompressed format for NONE as it require a new Compression strategy to be created in the future if we want float or int to support NONE.

gianm commented on processing/src/main/java/io/druid/segment/data/CompressedObjectStrategy.java:
```diff
@@ -85,6 +92,22 @@ public Compressor getCompressor()
+      @Override
+      public Compressor getCompressor()
+      {
+        return UncompressedCompressor.defaultCompressor;
```
Similar comment to getDecompressor

gianm commented on processing/src/main/java/io/druid/segment/data/CompressionFactory.java:
```diff
@@ -0,0 +1,275 @@
+  // 126 is the value here since -2 - 126 = -128, which is the lowest byte value
+  private static int FLAG_VALUE = 126;
+
+  public static boolean hasFlag(byte strategyId) {
```
minor: IMO has/put/removeFlag would be more clearly named hasEncodingFlag, setEncodingFlag, clearEncodingFlag.

acslk replied:
changed the names

gianm commented on processing/src/main/java/io/druid/segment/data/CompressionFactory.java:
```diff
@@ -0,0 +1,275 @@
+      return idMap.get(id);
+    }
+
+    public static LongEncoding[] testValues() {
```
This method needs a comment explaining what it's for.

acslk replied:
removed this method after separating EncodingStrategy and EncodingFormat

gianm commented on processing/src/main/java/io/druid/segment/data/CompressionFactory.java:
```diff
@@ -0,0 +1,275 @@
+ * <p>
+ * Values
+ */
+public abstract class CompressionFactory
```
I think there are no instances of this class, right? If not, drop the "abstract", there's no need for it. You can also signify that this is really just a bag of methods by having a private constructor, which disables the ability for anyone to instantiate it.

example:

```java
private CompressionFactory() {
  // No instantiation
}
```

acslk replied:
Changed class to non abstract and added private default constructor

gianm commented on processing/src/main/java/io/druid/segment/data/CompressionFactory.java:
```diff
@@ -0,0 +1,275 @@
+
+  public static final int MAX_TABLE_SIZE = 256;
+
+  /**
```
This is a javadoc style comment (`/**`) but it is not actually attached to anything. It should just be a regular multiline comment (`/*`).

acslk replied:
Changed comment style to regular comment

gianm commented on processing/src/main/java/io/druid/segment/data/CompressionFactory.java:
```diff
@@ -0,0 +1,275 @@
+    if (encoding == LongEncoding.TABLE || encoding == LongEncoding.DELTA) {
+      return new IntermediateLongSupplierSerializer(ioPeon, filenameBase, order, strategy);
+    }
+    if (strategy == CompressedObjectStrategy.CompressionStrategy.NONE) {
```
why not `else if` :P

gianm commented on processing/src/main/java/io/druid/segment/data/CompressedLongsIndexedSupplier.java:
```diff
@@ -73,302 +58,57 @@ public int size()
+    channel.write(buffer.asReadOnlyBuffer());
   }
 
   public CompressedLongsIndexedSupplier convertByteOrder(ByteOrder order)
```
It looks like this is only called by MetricHolder.convertByteOrder, which nothing calls. let's just remove both.

acslk replied:
removed both methods

gianm commented on processing/src/main/java/io/druid/segment/data/LongSupplierSerializer.java:
```diff
@@ -0,0 +1,38 @@
+
+/**
+ */
+public interface LongSupplierSerializer extends Closeable
```
This class should get some more javadocs that answer stuff like:

- what does `closeAndConsolidate` mean and how is that different from `close`?
- what does `writeToChannel` do, and should that be called before or after closing?
- when should `add` be called relative to other methods?
- when should `size` and `getSerializedSize` be called relative to other methods?
- what units does `size` return (bytes, number of longs)?

gianm replied:
Actually given that this is just copied from some existing code, it's okay if you only want to focus on the most confusing question, which for me is the difference between `closeAndConsolidate` and `close`.

acslk replied:
closeAndConsolidate is basically close + write everything to a give outputStream, and is used by IndexMerger. close by itself is used together with writeToChannel, which write everything to a give channel,  in IndexMergerv9, and close is always called before writeToChannel.

gianm commented on docs/content/development/extensions-core/kafka-ingestion.md:
```diff
@@ -128,9 +128,10 @@ The tuningConfig is optional and default parameters will be used if no tuningCon
+|bitmap|Object|Compression format for bitmap indexes. Should be a JSON object; see below for options.|no (defaults to Concise)|
+|dimensionCompression|String|Compression format for dimension columns. Choose from `LZ4`, `LZF`, `uncompressed`, or `none`. Note that `uncompressed` is deprecated and should only be used for backward compatibility, for newer versions, use `none` instead. |no (default == `LZ4`)|
+|metricCompression|String|Compression format for metric columns. Choose from `LZ4`, `LZF`, `uncompressed`, or `none`.|no (default == `LZ4`)|
+|longEncoding|String|Encoding format for metric and dimension columns with type long. Choose from `delta` or `longs`. `delta` encodes the values using offset or lookup table depending on column cardinality, and store them with variable size. `longs` stores the value as is with 8 bytes each.|no (default == `longs`)|
```
Judging from the code, looks like "delta" could also store as plain old longs encoding if neither delta nor table is appropriate. So it's really more of an "auto" encoding setting. The docs should reflect that.

gianm replied:
Is it possible to really call it "auto"? (i.e. have a different enum for "encodings that are used to write a column" vs "strategy for choosing an encoding")

gianm replied:
The fact that `getWriter` returns null for delta and table is also an indication that this split would be useful.

gianm commented on processing/src/main/java/io/druid/segment/data/LongsEncodingReader.java:
```diff
@@ -0,0 +1,65 @@
+  {
+    return new LongsEncodingReader(buffer.duplicate());
+  }
+}
```
Newline at EOF

acslk replied:
added newline

acslk commented:
Removed the reqInt, bytesLong, and timestamp files and now uses generated data. 
Added data sets includes: 
- enumerated (value from only a few selections, with probability heavily skewed toward one) 
- zipfLow (zipf distribution between 0 to 1000 with low exponent)
- zipfHigh (zipf distribution between 0 to 1000 with high exponent)
- uniform (uniform distribution between 0 to 1000)
- sequential (sequential values starting from a timestamp)

Generated files size in kb (5,000,000 values)
zipfLow-LZ4-DELTA: 6870
zipfLow-LZ4-LONGS: 9495
zipfLow-NONE-DELTA: 7324
zipfLow-NONE-LONGS: 39062
uniform-LZ4-DELTA: 7354
uniform-LZ4-LONGS: 11184
uniform-NONE-DELTA: 7324
uniform-NONE-LONGS: 39062
sequential-LZ4-DELTA: 14706
sequential-LZ4-LONGS: 19542
sequential-NONE-DELTA: 14648
sequential-NONE-LONGS: 39062
enumerate-LZ4-DELTA: 442
enumerate-LZ4-LONGS: 711
enumerate-NONE-DELTA: 2441
enumerate-NONE-LONGS: 39062
zipfHigh-LZ4-DELTA: 1201
zipfHigh-LZ4-LONGS: 1784
zipfHigh-NONE-DELTA: 4884
zipfHigh-NONE-LONGS: 39062

More benchmarks:
```
CompressionBenchmark.readContinuous   enumerate     delta         lz4  avgt   25  30.648 ±  2.626  ms/op
CompressionBenchmark.readContinuous   enumerate     delta        none  avgt   25   7.650 ±  0.526  ms/op
CompressionBenchmark.readContinuous   enumerate     longs         lz4  avgt   25  19.126 ±  0.588  ms/op
CompressionBenchmark.readContinuous   enumerate     longs        none  avgt   25   2.885 ±  0.172  ms/op
CompressionBenchmark.readContinuous     zipfLow     delta         lz4  avgt   25  36.759 ±  2.715  ms/op
CompressionBenchmark.readContinuous     zipfLow     delta        none  avgt   25  11.347 ±  0.733  ms/op
CompressionBenchmark.readContinuous     zipfLow     longs         lz4  avgt   25  32.491 ±  2.123  ms/op
CompressionBenchmark.readContinuous     zipfLow     longs        none  avgt   25   2.991 ±  0.267  ms/op
CompressionBenchmark.readContinuous    zipfHigh     delta         lz4  avgt   25  26.057 ±  1.361  ms/op
CompressionBenchmark.readContinuous    zipfHigh     delta        none  avgt   25   3.545 ±  0.227  ms/op
CompressionBenchmark.readContinuous    zipfHigh     longs         lz4  avgt   25  21.950 ±  1.035  ms/op
CompressionBenchmark.readContinuous    zipfHigh     longs        none  avgt   25   2.883 ±  0.175  ms/op
CompressionBenchmark.readContinuous  sequential     delta         lz4  avgt   25  28.658 ±  0.567  ms/op
CompressionBenchmark.readContinuous  sequential     delta        none  avgt   25   7.557 ±  0.495  ms/op
CompressionBenchmark.readContinuous  sequential     longs         lz4  avgt   25  53.250 ±  1.897  ms/op
CompressionBenchmark.readContinuous  sequential     longs        none  avgt   25   3.117 ±  0.270  ms/op
CompressionBenchmark.readContinuous     uniform     delta         lz4  avgt   25  32.909 ±  2.348  ms/op
CompressionBenchmark.readContinuous     uniform     delta        none  avgt   25  20.803 ± 11.509  ms/op
CompressionBenchmark.readContinuous     uniform     longs         lz4  avgt   25  30.252 ±  0.674  ms/op
CompressionBenchmark.readContinuous     uniform     longs        none  avgt   25   2.944 ±  0.130  ms/op
CompressionBenchmark.readSkipping     enumerate     delta         lz4  avgt   25   1.462 ±  0.091  ms/op
CompressionBenchmark.readSkipping     enumerate     delta        none  avgt   25   0.016 ±  0.001  ms/op
CompressionBenchmark.readSkipping     enumerate     longs         lz4  avgt   25   5.476 ±  1.154  ms/op
CompressionBenchmark.readSkipping     enumerate     longs        none  avgt   25   0.055 ±  0.001  ms/op
CompressionBenchmark.readSkipping       zipfLow     delta         lz4  avgt   25   4.786 ±  0.153  ms/op
CompressionBenchmark.readSkipping       zipfLow     delta        none  avgt   25   0.034 ±  0.001  ms/op
CompressionBenchmark.readSkipping       zipfLow     longs         lz4  avgt   25  17.017 ±  0.565  ms/op
CompressionBenchmark.readSkipping       zipfLow     longs        none  avgt   25   0.054 ±  0.001  ms/op
CompressionBenchmark.readSkipping      zipfHigh     delta         lz4  avgt   25   2.745 ±  0.078  ms/op
CompressionBenchmark.readSkipping      zipfHigh     delta        none  avgt   25   0.023 ±  0.001  ms/op
CompressionBenchmark.readSkipping      zipfHigh     longs         lz4  avgt   25   7.876 ±  0.972  ms/op
CompressionBenchmark.readSkipping      zipfHigh     longs        none  avgt   25   0.055 ±  0.001  ms/op
CompressionBenchmark.readSkipping    sequential     delta         lz4  avgt   25   1.239 ±  0.374  ms/op
CompressionBenchmark.readSkipping    sequential     delta        none  avgt   25   0.048 ±  0.001  ms/op
CompressionBenchmark.readSkipping    sequential     longs         lz4  avgt   25  36.913 ±  0.918  ms/op
CompressionBenchmark.readSkipping    sequential     longs        none  avgt   25   0.054 ±  0.001  ms/op
CompressionBenchmark.readSkipping       uniform     delta         lz4  avgt   25   0.660 ±  0.218  ms/op
CompressionBenchmark.readSkipping       uniform     delta        none  avgt   25   0.033 ±  0.001  ms/op
CompressionBenchmark.readSkipping       uniform     longs         lz4  avgt   25  17.497 ±  0.561  ms/op
CompressionBenchmark.readSkipping       uniform     longs        none  avgt   25   0.054 ±  0.001  ms/op
```

gianm commented on processing/src/main/java/io/druid/segment/data/LongsEncodingReader.java:
```diff
@@ -0,0 +1,65 @@
+import java.nio.ByteOrder;
+import java.nio.LongBuffer;
+
+public class LongsEncodingReader implements CompressionFactory.LongEncodingReader
```
This should really be called LongsLongEncodingReader. It's weird but it's more consistent :)

acslk replied:
Changed all LongEncodingReaders and LongEncodingWriters to have LongEncoding in their name instead of Encoding

gianm commented on processing/src/main/java/io/druid/segment/data/LongsEncodingReader.java:
```diff
@@ -0,0 +1,65 @@
+    this.buffer = fromBuffer.asReadOnlyBuffer().order(order).asLongBuffer();
+  }
+
+  public LongsEncodingReader(LongBuffer buffer)
```
This could be private.

gianm commented on processing/src/main/java/io/druid/segment/data/LongsEncodingWriter.java:
```diff
@@ -0,0 +1,96 @@
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+public class LongsEncodingWriter implements CompressionFactory.LongEncodingWriter
```
This should be called LongsLongEncodingWriter

gianm commented on processing/src/main/java/io/druid/segment/data/DeltaEncodingReader.java:
```diff
@@ -0,0 +1,79 @@
+
+import java.nio.ByteBuffer;
+
+public class DeltaEncodingReader implements CompressionFactory.LongEncodingReader
```
Similar comment for Table*

gianm commented on processing/src/main/java/io/druid/segment/data/DeltaEncodingWriter.java:
```diff
@@ -0,0 +1,88 @@
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+
+public class DeltaEncodingWriter implements CompressionFactory.LongEncodingWriter
```
This should be DeltaLongEncodingWriter

gianm commented:
@acslk sweet, those benchmarks look nice. The general theme seems to be,

- longs+lz4 is the current default encoding
- delta+lz4 is always smallest and is roughly similar speed to longs+lz4; maybe this'd be a nice default in the future
- delta+none size varies quite a bit relative to longs+lz4, but is always faster
- longs+none is silly 😄 

Could you put those in the top comment too for ease of reference?

acslk commented:
I've added some basic float encoding stuff that supports non block based float compression in another branch, and hoping to get it merged together with this branch. 
https://github.com/acslk/druid/compare/feature-compress...acslk:feature-floatcompress?expand=1

The reason is that Long metric and Float metric share the same CompressionStrategy in IndexSpec, and updating only Long would cause Long to compress NONE strategy in one way while Float compress it in another way, which is awkward for future Float encoding implementation.

The float changes are mostly symmetrical to changes in this PR.

gianm commented:
@acslk hmm, it probably makes sense to build real "none" compression support for floats in this PR, since that'll remove the awkwardness in IndexMerger / IndexMergerV9, and make the migration path better later. Good call.

But I don't think we need the floatEncoding stuff - generally I'd prefer to add that interface when we actually need it. Right now we don't since there's only one float encoding (and I'm not creative enough to think of a different one we might want to do in the near future). So if removing that would simplify the float stuff, let's remove it.

gianm commented on processing/src/main/java/io/druid/segment/data/DeltaLongEncodingReader.java:
```diff
@@ -0,0 +1,79 @@
+    if (version == CompressionFactory.DELTA_ENCODING_VERSION) {
+      base = buffer.getLong();
+      bitsPerValue = buffer.getInt();
+      fromBuffer.position(buffer.position());
```
Why does this need to set the position of `fromBuffer`?

acslk replied:
The encoding readers take the encoding header from the input byte buffer, and set the position to the content after the encoding header for LayoutSupplier to use.

gianm commented on processing/src/main/java/io/druid/segment/data/DeltaLongEncodingWriter.java:
```diff
@@ -0,0 +1,88 @@
+  public DeltaLongEncodingWriter(long base, long delta)
+  {
+    this.base = base;
+    this.bitsPerValue = VSizeLongSerde.getBitsForMax(delta + 1);
```
What if delta is `Long.MaxValue`? (let's say a column has values ranging from 0 to Long.MaxValue). Then we'll be calling `VSizeLongSerde.getBitsForMax(-1)`.

Is there a reason this needs to be `delta + 1` instead of `delta`?

gianm replied:
Would be good to have a test that hits that case (positive longs ranging from 0 to max).

acslk replied:
getBitsForMax(n) returns bits required to encode at most n unique values. delta is the difference between the smallest and largest value, which means there would be delta + 1 unique values since smallest and largest value are both included in the set. 

For the delta = Long.MaxValue case it would cause an exception, and I can change the strategy to select Longs format for this case. Also going to add a test with exactly max long delta.

gianm commented on processing/src/main/java/io/druid/segment/data/DeltaLongEncodingWriter.java:
```diff
@@ -0,0 +1,88 @@
+      serializer.close();
+    }
+  }
+}
```
Newline @ EOF

gianm commented on processing/src/main/java/io/druid/segment/data/TableLongEncodingReader.java:
```diff
@@ -0,0 +1,85 @@
+      for (int i = 0; i < tableSize; i++) {
+        table[i] = buffer.getLong();
+      }
+      fromBuffer.position(buffer.position());
```
Why do we need to update the position of `fromBuffer`? (perhaps a comment is in order)

acslk replied:
some comment as above, all encoding readers would update the position to after encoding header

gianm commented on processing/src/main/java/io/druid/segment/data/TableLongEncodingWriter.java:
```diff
@@ -0,0 +1,97 @@
+  {
+    return VSizeLongSerde.getSerializedSize(bitsPerValue, values);
+  }
+}
```
Newline

gianm commented:
@acslk Okay, whew, finally read through the patch :)

I wrote some comments on specific areas but the general structure looks good to me. The benchmarks and storage numbers look good too!

Other than that, like I said in the previous comment, I do think it makes sense to include "none" compression for floats, minus the probably unnecessary floatEncoding stuff.

Anyone else have a chance to take a look at this now that it's been edited a bunch? @xvrl @jon-wei

jon-wei commented:
@acslk @gianm reviewing again

acslk commented:
Added the float support for "none" compression strategy. It uses the same BlockLayout/EntireLayout pattern as long compression, but does not have any encoding strategy/format.

xvrl commented:
@acslk I'm a bit puzzled by some of the benchmark numbers. `longs + lz4` seems to be really standing out in the readSkipping benchmarks compared to `delta + lz4`. The difference is much less pronounced in readContinous. I wonder it it would make sense to understand what's causing this, especially give that the compressed sizes are not that much different.

@gianm As far as making `delta+lz4` the default in the future, it still seems to be up to 3x slower compared to `longs + lz4` for some full column scans, so we might want to experiment with real-world data before making that decision.

Overall I'm on board with this PR, I haven't had a chance to read the code much yet, but the approach is the description seems reasonable.

jon-wei commented on processing/src/main/java/io/druid/segment/IndexSpec.java:
```diff
@@ -79,30 +113,39 @@ public IndexSpec()
       @JsonProperty("bitmap") BitmapSerdeFactory bitmapSerdeFactory,
       @JsonProperty("dimensionCompression") String dimensionCompression,
-      @JsonProperty("metricCompression") String metricCompression
+      @JsonProperty("metricCompression") String metricCompression,
```
I think the implementation of IndexSpec could be simplified by using the Enums directly in the constructor, instead of Strings, jackson can serialize/deserialize those to/from strings.

SegmentMetadataQuery.AnalysisType is an example using @JsonValue and @JsonCreator for that purpose

acslk replied:
it seems that jackson cannot convert input string to uppercase for the enum, so serialize/deserialize need to take string and apply conversion.

acslk replied:
Nevermind, added the toString and fromString for the enums and replaced the string with enum values. There seems to be a special handle for Uncompressed strategy for dimension compression, where getStrategy returns null for uncompressed. The caller handles the null as indicator to not use compressedIndexedInt, I think it's doing something similar to none strategy in this PR for performance reasons. I removed the null return value and did the special handling for UNCOMPRESSED strategy instead, this syntax change should not affect anything.

jon-wei commented on processing/src/main/java/io/druid/segment/data/VSizeLongSerde.java:
```diff
@@ -0,0 +1,634 @@
+
+  public static int getSerializedSize(int longSize, int numValues)
+  {
+    return (longSize * numValues + 7) / 8 + 4;
```
can you add a comment indicating how the serialized size is calculated? what do the 7 and +4 represent?

acslk replied:
Added comment for the calculation

jon-wei commented on processing/src/main/java/io/druid/segment/data/VSizeLongSerde.java:
```diff
@@ -0,0 +1,634 @@
+public class VSizeLongSerde
+{
+
+  public static final int SUPPORTED_SIZE[] = {1, 2, 4, 8, 12, 16, 20, 24, 32, 40, 48, 56, 64};
```
how come 28, 36, 44, 52, and 60 aren't supported byte sizes?

acslk replied:
I followed the same supported size for lucene, I believe the rational is that bit sizes that are not multiples of 8 takes more operations to serialize/deserialize, so performance is worse. The larger values not divisible by 8 doesn't offer a significant size reduction to next multiple of 8 compare to smaller values. For instance, 12 offers 25% size reduction to 16, while 36 only has 10% size reduction to 40, so it's not worth the performance difference.

jon-wei replied:
cool, can you document that reasoning in a comment?

jon-wei commented on benchmarks/src/main/java/io/druid/benchmark/FloatCompressionBenchmark.java:
```diff
@@ -0,0 +1,106 @@
+  {
+    sum = 0;
+    for (int i = 0; i < count; i++) {
+      sum += indexedFloats.get(i);
```
The JMH benchmarks should return the calculated value or feed them into a blackhole, see:

"Dead Code" section:
http://java-performance.info/jmh/  

http://hg.openjdk.java.net/code-tools/jmh/file/397891feda1b/jmh-samples/src/main/java/org/openjdk/jmh/samples/JMHSample_09_Blackholes.java

acslk replied:
Used blackhole to consume sum instead of printing it in teardown.

jon-wei commented:
done reviewing the new changes, had a few comments but looks good overall

acslk commented:
@xvrl After some investigation, I believe the reason for the discrepancy is due to performance of different components. For block based compression layout, the total time for reading column is composed of : copy byte buffer block + decompression + read values from byte buffer. 

`longs + lz4` takes more time on the decompression part, while `delta + lz4` take more time on the read value part. For readContinuous, the two differences kinda cancel out, making the performance similar. For readSkipping, the read value part is basically gone with so few reads, while the decompression part remains around the same, making `longs + lz4` take much longer.

I did some benchmark using the uncompressed strategy, which goes through the same process as lz4 for copy block + read value, with the only difference being no decompression step. The difference in time between lz4 and uncompressed should be the time for decompression. From the benchmarks below, it seems that readContinuous and readSkipping decompression time matches.

Note delta strategy is renamed to auto since the strategy could use delta, table, or longs depending on data cardinality and maximum offset.
```
enumerate
readContinuous
 auto           lz4  10  29.033 ± 1.111  ms/op
 auto  uncompressed  10  28.591 ± 0.911  ms/op
longs           lz4  10  20.878 ± 1.856  ms/op
longs  uncompressed  10  17.101 ± 0.358  ms/op
readSkipping
 auto           lz4  10   1.459 ± 0.069  ms/op
 auto  uncompressed  10   0.236 ± 0.123  ms/op
longs           lz4  10   5.889 ± 2.222  ms/op
longs  uncompressed  10   3.327 ± 1.929  ms/op

zipfLow
readContinuous
 auto           lz4  10  35.505 ± 0.744  ms/op
 auto  uncompressed  10  31.311 ± 1.022  ms/op
longs           lz4  10  33.904 ± 2.328  ms/op
longs  uncompressed  10  17.404 ± 0.690  ms/op
readSkipping
 auto           lz4  10   4.941 ± 0.360  ms/op
 auto  uncompressed  10   0.522 ± 0.141  ms/op
longs           lz4  10  18.387 ± 1.321  ms/op
longs  uncompressed  10   3.390 ± 2.403  ms/op


zipfHigh
readContinuous
 auto           lz4  10  26.399 ± 1.428  ms/op
 auto  uncompressed  10  23.593 ± 0.667  ms/op
longs           lz4  10  23.355 ± 1.868  ms/op
longs  uncompressed  10  17.335 ± 0.796  ms/op
readSkipping
 auto           lz4  10   2.914 ± 0.259  ms/op
 auto  uncompressed  10   0.383 ± 0.142  ms/op
longs           lz4  10   8.015 ± 0.590  ms/op
longs  uncompressed  10   3.406 ± 2.738  ms/op

sequential
readContinuous
 auto           lz4  10  29.491 ± 0.951  ms/op
 auto  uncompressed  10  28.636 ± 0.351  ms/op
longs           lz4  10  54.110 ± 0.894  ms/op
longs  uncompressed  10  17.035 ± 0.249  ms/op
readSkipping
 auto           lz4  10   1.335 ± 0.735  ms/op
 auto  uncompressed  10   1.079 ± 0.282  ms/op
longs           lz4  10  39.549 ± 0.544  ms/op
longs  uncompressed  10   3.472 ± 2.714  ms/op

uniform
readContinuous
 auto           lz4  10  31.522 ± 1.021  ms/op
 auto  uncompressed  10  33.260 ± 1.238  ms/op
longs           lz4  10  32.292 ± 1.532  ms/op
longs  uncompressed  10  17.211 ± 0.955  ms/op
readSkipping
 auto           lz4  10   0.746 ± 0.518  ms/op
 auto  uncompressed  10   0.528 ± 0.176  ms/op
longs           lz4  10  18.018 ± 1.105  ms/op
longs  uncompressed  10   3.302 ± 2.018  ms/op
```
I'm not sure why lz4 decompression is faster for auto compared to longs even though the compressed size are similar. Perhaps less work is done for auto decompression since the decompressed size is smaller.

jon-wei commented on benchmarks/src/main/java/io/druid/benchmark/LongCompressionBenchmark.java:
```diff
@@ -0,0 +1,108 @@
+// Run LongCompressionBenchmarkFileGenerator to generate the required files before running this benchmark
+
+@State(Scope.Benchmark)
+@Fork(value = 1)
```
Let's make this:
```
@Fork(jvmArgsPrepend = "-server", value = 1)
```

so that the JVM runs in "server" mode:
http://www.oracle.com/technetwork/java/whitepaper-135217.html#2

gianm commented:
@xvrl yes, that's a good point, looks like @acslk findings indicate that auto+lz4 generally decompresses faster than longs+lz4 but individual values are then read out slower (which explains why full scans are hit-or-miss but filtered scans are generally faster in auto+lz4's favor).

I just read over the more recent changes (the float compression stuff and the introduction of "auto" encoding strategy) and they look good to me. We've also been running this on our test cluster for a couple of weeks now and it has been working out there.

👍  from me, any one else want to take one more look? @jon-wei @xvrl

jon-wei commented:
LGTM, :+1:

gianm commented:
Looks like we have a few +1 and no objections, so I'll merge this as discussed in the dev sync today. thanks everyone!

drcrallen commented on processing/src/main/java/io/druid/segment/data/GenericIndexedWriter.java:
```diff
@@ -132,20 +134,20 @@ public long getSerializedSize()
+  public ByteSource combineStreams()
   {
-    return ByteStreams.join(
+    return ByteSource.concat(
```
This broke batch indexing that uses Guava pre-15

leventov commented on processing/src/main/java/io/druid/segment/IndexSpec.java:
```diff
@@ -41,36 +40,34 @@
-  private final String metricCompression;
+  private final CompressedObjectStrategy.CompressionStrategy dimensionCompression;
+  private final CompressedObjectStrategy.CompressionStrategy metricCompression;
+  private final CompressionFactory.LongEncodingStrategy longEncoding;
```
Shouldn't `longEncoding` be checked in `IndexSpec.equals()` and `hashCode()`?

gianm replied:
Yes, it should.

leventov replied:
Fixed that here: #4353
